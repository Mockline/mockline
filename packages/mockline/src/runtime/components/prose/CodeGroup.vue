<script setup lang="ts">
import { useSlots, computed } from 'vue'

defineOptions({
  inheritAttrs: false
})

type CodeGroupProps = {
  class?: any
}

const props = defineProps<CodeGroupProps>()

const slots = useSlots()

const selectedIndex = ref(0)
defineExpose({ selectedIndex })

function transformSlot(slot: any, index: number): any {
  if (typeof slot.type === 'symbol') {
    return slot.children?.map(transformSlot)
  }

  return {
    label: slot.props?.filename || slot.props?.label || `${index}`,
    icon: slot.props?.icon,
    component: slot
  }
}

const tabs = computed(() => slots.default?.()?.flatMap(transformSlot).filter(Boolean) || [])

const selectedTab = computed(() => tabs.value.find((_, index) => index === selectedIndex.value))
</script>

<template>
  <div class="relative my-5 [&>div:last-child]:!static [&>div:last-child]:!my-0">
    <div class="flex items-center gap-1 overflow-hidden rounded-t-md bg-neutral-100 p-2 dark:bg-neutral-800">
      <button
        v-for="(tab, index) in tabs"
        :key="index"
        tabindex="-1"
        :class="['flex items-center gap-1.5 rounded-md px-2 py-1.5 text-sm text-neutral-950 focus:outline-none dark:text-neutral-100',
                 selectedIndex === index ? 'bg-neutral-200 dark:bg-neutral-700' : 'hover:bg-neutral-300 dark:hover:bg-neutral-700']"
        @click="selectedIndex = index"
      >
        <CodeIcon :icon="tab.icon" :filename="tab.label" />
        <span>{{ tab.label }}</span>
      </button>
    </div>

    <component :is="selectedTab?.component" :key="selectedIndex" hide-header />
  </div>
</template>
