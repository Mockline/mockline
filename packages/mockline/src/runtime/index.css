@import "tailwindcss";
@import "assets/css/animation.css";

@variant dark (&:where(.dark, .dark *));

@plugin "@tailwindcss/typography";

@theme {
  --color-neutral: var(--ui-bg);
  --color-neutral-elevated: var(--ui-bg-elevated);
  --color-border: var(--ui-border);
}

@layer base {
  :root {
    color-scheme: light;

    --ui-text-dimmed: var(--ui-color-neutral-400);
    --ui-text-muted: var(--ui-color-neutral-500);
    --ui-text-toned: var(--ui-color-neutral-600);
    --ui-text: var(--ui-color-neutral-700);
    --ui-text-highlighted: var(--ui-color-neutral-900);

    --ui-bg: var(--color-white);
    --ui-bg-elevated: var(--ui-color-neutral-100);
    --ui-bg-accented: var(--ui-color-neutral-200);
    --ui-bg-inverted: var(--ui-color-neutral-900);

    --ui-border: var(--ui-color-neutral-200);
    --ui-border-accented: var(--ui-color-neutral-300);
    --ui-border-inverted: var(--ui-color-neutral-900);

    --ui-radius: var(--radius);

    /*--ui-slate: var(--color-slate-500);
    --ui-gray: var(--color-gray-500);
    --ui-zinc: var(--color-zinc-500);
    --ui-stone: var(--color-stone-500);
    --ui-neutral: var(--color-neutral-500);
    --ui-red: var(--color-red-500);
    --ui-orange: var(--color-orange-500);
    --ui-amber: var(--color-amber-500);
    --ui-yellow: var(--color-yellow-500);
    --ui-lime: var(--color-lime-500);
    --ui-green: var(--color-green-500);
    --ui-emerald: var(--color-emerald-500);
    --ui-teal: var(--color-teal-500);
    --ui-cyan: var(--color-cyan-500);
    --ui-sky: var(--color-sky-500);
    --ui-blue: var(--color-blue-500);
    --ui-indigo: var(--color-indigo-500);
    --ui-violet: var(--color-violet-500);
    --ui-purple: var(--color-purple-500);
    --ui-fuchsia: var(--color-fuchsia-500);
    --ui-pink: var(--color-pink-500);
    --ui-rose: var(--color-rose-500);*/
  }

  .dark {
    color-scheme: dark;

    --ui-text-dimmed: var(--ui-color-neutral-500);
    --ui-text-muted: var(--ui-color-neutral-400);
    --ui-text-toned: var(--ui-color-neutral-300);
    --ui-text: var(--ui-color-neutral-200);
    --ui-text-highlighted: var(--color-white);

    --ui-bg: var(--ui-color-neutral-900);
    --ui-bg-elevated: var(--ui-color-neutral-800);
    --ui-bg-accented: var(--ui-color-neutral-700);
    --ui-bg-inverted: var(--color-white);

    --ui-border: var(--ui-color-neutral-800);
    --ui-border-accented: var(--ui-color-neutral-700);
    --ui-border-inverted: var(--color-white);
  }
}

