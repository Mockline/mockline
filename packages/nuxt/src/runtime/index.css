@import "@mockline/themes";
@import "./mockline.css";
