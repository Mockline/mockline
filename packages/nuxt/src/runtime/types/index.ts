// Elements
export * from '../components/Button.vue'
export * from '../components/NavMenu.vue'

// Layout
export * from '../components/App.vue'
export * from '../components/Aside.vue'
export * from '../components/Card.vue'
export * from '../components/Header.vue'
export * from '../components/Main.vue'
export * from '../components/Page.vue'
export * from '../components/PageBody.vue'
export * from '../components/PageHeader.vue'

// Settings
export * from '../components/ThemeToggle.vue'

// Overlays
export * from '../components/Toasts.vue'

// Content
export * from '../components/content/ContentNavigationTree.vue'
export * from '../components/content/ContentSurround.vue'
export * from '../components/content/ContentToc.vue'
export * from '../components/content/ContentTocLinks.vue'

// Global
export * from '../components/prose/Callout.vue'
export * from '../components/prose/CodeGroup.vue'
export * from '../components/prose/ProseCode.vue'
export * from '../components/prose/CodeButton.vue'
