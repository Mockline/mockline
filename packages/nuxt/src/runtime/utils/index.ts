export function transformSlot(slot: any, index: number): any {
  if (typeof slot.type === 'symbol') {
    return slot.children?.map(transformSlot)
  }
  return {
    label: slot.props?.filename || slot.props?.label || `${index}`,
    icon: slot.props?.icon,
    component: slot
  }
}

export * from './useComponent'
