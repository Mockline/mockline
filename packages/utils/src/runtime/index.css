@import "tailwindcss";

@variant dark (&:where(.dark, .dark *));

@plugin "@tailwindcss/typography";
