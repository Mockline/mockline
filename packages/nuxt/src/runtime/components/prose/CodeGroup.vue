<script setup lang="ts">
import { TabsRoot, TabsList, TabsIndicator, TabsTrigger, TabsContent } from 'reka-ui'
import { useSlots, computed, onMounted, watch } from 'vue'
import { tv } from 'tailwind-variants'
import { transformSlot } from '../../utils'
import CodeIcon from './CodeIcon.vue'
import { useState } from '#imports'

const theme = tv({
  slots: {
    root: 'relative group [&>div:not(:first-child)]:!my-0 [&>div:not(:first-child)]:!static my-5',
    list: 'relative flex items-center gap-1 border border-[var(--ui-border-muted)] bg-[var(--ui-bg)] border-b-0 rounded-t-[calc(var(--ui-radius)*1.5)] overflow-x-auto p-2',
    indicator: 'absolute left-0 inset-y-2 w-[var(--reka-tabs-indicator-size)] translate-x-[var(--reka-tabs-indicator-position)] transition-[translate,width] duration-200 bg-[var(--ui-bg-elevated)] rounded-[calc(var(--ui-radius)*1.5)] shadow-xs',
    trigger: ['relative inline-flex items-center gap-1.5 text-[var(--ui-text)] data-[state=active]:text-[var(--ui-text-highlighted)] hover:bg-[var(--ui-bg-elevated)]/50 px-2 py-1.5 text-sm rounded-[calc(var(--ui-radius)*1.5)] disabled:cursor-not-allowed disabled:opacity-75 focus-visible:ring-2 focus-visible:ring-inset focus-visible:ring-[var(--ui-primary)] focus:outline-none'],
    triggerIcon: 'size-4 shrink-0',
    triggerLabel: 'truncate'
  }
})

type CodeGroupProps = {
  defaultValue?: string
  class?: any
  sync?: string
}

const props = withDefaults(defineProps<CodeGroupProps>(), {
  defaultValue: '0'
})

const codeGroup = theme()

const slots = useSlots()

const model = defineModel<string>()

const items = computed(() => slots.default?.()?.flatMap(transformSlot).filter(Boolean) || [])

onMounted(() => {
  if (props.sync) {
    const syncKey = `code-group-${props.sync}`
    const syncValue = useState<string>(syncKey, () => localStorage.getItem(syncKey) as string)

    watch(syncValue, () => {
      if (!syncValue.value) return

      model.value = syncValue.value
    }, { immediate: true })

    watch(model, () => {
      if (!model.value) return

      syncValue.value = model.value
      localStorage.setItem(syncKey, model.value)
    })
  }
})
</script>

<template>
  <TabsRoot v-model="model" :default-value :class="codeGroup.root({ class: props.class })">
    <TabsList :class="codeGroup.list()">
      <TabsIndicator :class="codeGroup.indicator()" />

      <TabsTrigger v-for="(item, index) of items" :key="index" :value="String(index)" :class="codeGroup.trigger()">
        <CodeIcon :icon="item.icon" :filename="item.label" :class="codeGroup.triggerIcon()" />

        <span :class="codeGroup.triggerLabel()">{{ item.label }}</span>
      </TabsTrigger>
    </TabsList>

    <TabsContent v-for="(item, index) of items" :key="index" :value="String(index)" force-mount as-child>
      <component :is="item.component" hide-header />
    </TabsContent>
  </TabsRoot>
</template>
