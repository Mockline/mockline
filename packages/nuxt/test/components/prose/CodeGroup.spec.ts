import { describe, it, expect } from 'vitest'
import { renderSuspended } from '@nuxt/test-utils/runtime'
import CodeGroup from '../../src/runtime/components/prose/CodeGroup.vue'

describe('CodeGroup', () => {
  it('can mount component', async () => {
    const html = await renderSuspended(CodeGroup)
    expect(html.html()).toMatchSnapshot()
  })
})
