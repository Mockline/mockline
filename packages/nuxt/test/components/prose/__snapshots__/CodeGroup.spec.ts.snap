// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CodeGroup > renders basic case correctly 1`] = `
"<div dir="ltr" data-orientation="horizontal" class="pb-2 px-2 code-group relative group [&amp;>div:not(:first-child)]:!my-0 [&amp;>div:not(:first-child)]:!static my-5 rounded-base border border-[var(--color-border-primary)] bg-[var(--color-bg-primary)] dark">
  <div role="tablist" aria-orientation="horizontal" tabindex="-1" data-orientation="horizontal" dir="ltr" style="outline-color: none; outline-style: none; outline-width: initial;" class="relative flex items-center gap-1 overflow-x-auto py-2">
    <!---->
  </div>
</div>"
`;

exports[`CodeGroup > renders with defaultValue correctly 1`] = `
"<div dir="ltr" data-orientation="horizontal" class="pb-2 px-2 code-group relative group [&amp;>div:not(:first-child)]:!my-0 [&amp;>div:not(:first-child)]:!static my-5 rounded-base border border-[var(--color-border-primary)] bg-[var(--color-bg-primary)] dark">
  <div role="tablist" aria-orientation="horizontal" tabindex="-1" data-orientation="horizontal" dir="ltr" style="outline-color: none; outline-style: none; outline-width: initial;" class="relative flex items-center gap-1 overflow-x-auto py-2">
    <!---->
  </div>
</div>"
`;
